### PR TITLE
Feature/nexteuropa 8413

### PR DESCRIPTION
**Pull request by voidtek**
https://webgate.ec.europa.eu/CITnet/jira/browse/NEXTEUROPA-8413

### CODE REVIEW

File: profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php [hunk 1]
```diff
@@ -79,6 +79,13 @@ public function hookTokens($type, $tokens, array $data = array(), array $options
                 $this->watchdogTokenNotFound($data, $original);
               }
               break;
+
+            case 'bean':
+              $bean = bean_load($entity_id);
+              if (user_access('access user profiles')) {
```
Comment (cyberwolf) [inline profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php]: To me this seems to be a weird permission to check on in this context?

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 2]
```diff
@@ -1,18 +1,135 @@
...
-
 NextEuropa Webtools provides a bean for integrating with the Webtools service.
 
 Feature user documentation: https://webgate.ec.europa.eu/fpfis/wikis/display/MULTISITE/Webtools+feature
 
 Webtools service documentation: https://webgate.ec.europa.eu/fpfis/wikis/display/webtools/Webtools+documentation+-+Homepage
 
-Installation
-------------
+Table of content:
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: Table of content**s**

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 3]
```diff
@@ -1,18 +1,135 @@
...
+[Go to top](#table-of-content)
+
+# Configuration
+
+## Configure the Smartloader
+
+The only thing that needs to be configured is the url where the Smartloader is defined.
+This can be done as an administrative user on "/admin/config/services/webtools".
+Most sites can use the europa.eu smartloader (//europa.eu/webtools/load.js),
+but this can be configured to use an other one if needed.
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: another one

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 4]
```diff
@@ -1,18 +1,135 @@
...
+
+[Go to top](#table-of-content)
+
+# Configuration
+
+## Configure the Smartloader
+
+The only thing that needs to be configured is the url where the Smartloader is defined.
+This can be done as an administrative user on "/admin/config/services/webtools".
+Most sites can use the europa.eu smartloader (//europa.eu/webtools/load.js),
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: Are the double leading slashes (//) intentionally?

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 5]
```diff
@@ -1,18 +1,135 @@
...
+    - External link allows user to use an external JavaScript file.
+- Click on 'Save'
+- You are automatically redirected to the result
+
+[Go to top](#table-of-content)
+
+## How to: use the webtools block(s)
+
+The created blocks are listed at "admin/content/blocks".
+These blocks can be uses as any other block in Drupal.
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: can be use**d**

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 6]
```diff
@@ -1,18 +1,135 @@
...
+You can add them to the page with the block system "admin/structure/blocks" -
+or via context "admin/structure/context" or via any other system to place blocks
+in Drupal.
+
+Note that the blocks created are named Beans, as the Webtools feature is based
+on beans module.
+
+[Go to top](#table-of-content)
+
+## How to: insert a existing webtools block into a content
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: into content

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 7]
```diff
@@ -1,18 +1,135 @@
...
+
+As an authorized user, edit the content and go to the CKEditor:
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: I'd call it WYSIWYG editor here, end users do not necessarily know the product CKEditor.

File: profiles/common/modules/features/nexteuropa_webtools/README.md [hunk 8]
```diff
@@ -1,18 +1,135 @@
...
+Steps:
+- Click "Insert internal content"
+- Go in the tab "Insert internal blocks"
+- Select a block and click the link "Default" in the column "Insert as"
+
+[Go to top](#table-of-content)
+
+# Testing
+
+The Behat test exist in the file "/tests/features/webtools.feature".
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: exist**s** or "is located"

File: tests/features/webtools.feature [hunk 9]
```diff
@@ -0,0 +1,44 @@
+@api
+Feature: Webtools as administrator
+  In order to manage Webtools on the website
```
Comment (cyberwolf) [inline tests/features/webtools.feature]: This is not the goal the business has. Webtools is just the means, not the goal. What does webtools offer as an added value to the website?

File: tests/features/webtools.feature [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  Background:
+    Given these modules are enabled
+      | modules             |
+      | nexteuropa_webtools |
+    And a valid Smartload Url has been configured
+    And I am logged in as a user with the 'administrator' role
+
+  @api
+  Scenario: Create and delete a block 'Map'
```
Comment (cyberwolf) [inline tests/features/webtools.feature]: Ideally these would be 2 separate scenarios.

File: tests/src/Context/WebtoolsContext.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * Create a Webtools blocks.
+   *
+   * @param string $name
+   *   Name of the block webtools.
+   *
+   * @Given a map webtools :name exists
+   *
+   * @Then I create a new map webtools :name
```
Comment (cyberwolf) [inline tests/src/Context/WebtoolsContext.php]: Shouldn't this be @When ?

File: tests/src/Context/WebtoolsContext.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  /**
+   * Constructs a new WebtoolsContext.
+   *
+   * @param string $smartloadurl
+   *   The Smarloader Protocol-Relative URL.
+   */
+  public function __construct($smartloadurl) {
+    // @todo continuousphp insert the right $url
```
Comment (cyberwolf) [inline tests/src/Context/WebtoolsContext.php]: Can we remove the todo?
Which URL should be used on continuousphp instead?

File: tests/src/Context/WebtoolsContext.php [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    // $this->smartloadUrl = $smartloadurl;
+    $this->smartloadUrl = "http://europa.eu/webtools/load.js";
+  }
+
+  /**
+   * Configures the Smarloader URL for the nexteuropa_webtools module.
```
Comment (cyberwolf) [inline tests/src/Context/WebtoolsContext.php]: Smar**t**loader

File: tests/src/Context/WebtoolsContext.php [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+   *
+   * @Given a valid Smartload Url has been configured
+   */
+  public function aValidSmartloadUrlHasBeenConfigured() {
+    $this->variableContext->setVariable('nexteuropa_webtools_smartloader_prurl', $this->smartloadurl);
+  }
+
+  /**
+   * Create a Webtools blocks.
```
Comment (cyberwolf) [inline tests/src/Context/WebtoolsContext.php]: a ... block.

File: profiles/common/modules/features/nexteuropa_webtools/nexteuropa_webtools.module [hunk 15]
```diff
@@ -21,6 +28,14 @@ function nexteuropa_webtools_menu() {
     'file' => 'nexteuropa_webtools.admin.inc',
     'type' => MENU_NORMAL_ITEM,
   );
+
+  $items['nexteuropa/webtools-ckeditor/%'] = array(
+    'page callback' => 'nexteuropa_webtools_ajax_callback',
+    'page arguments' => array(2, 3),
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/nexteuropa_webtools.module]: nexteuropa_webtools_ajax_callback() only accepts one argument, probably 3 is not applicable here?

File: profiles/common/modules/features/nexteuropa_webtools/README.md (hunk 4 above)
Comment (voidtek) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: yes

File: tests/features/webtools.feature (hunk 10 above)
Comment (voidtek) [inline tests/features/webtools.feature]: This two cases are very simple and I think that it's not necessary to separate them.

File: profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php (hunk 1 above)
Comment (voidtek) [inline profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php]: What is your proposition?

File: profiles/common/modules/features/nexteuropa_webtools/README.md (hunk 4 above)
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_webtools/README.md]: Ok thanks, I only found that out after manually playing with the feature.

File: profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php (hunk 1 above)
Comment (cyberwolf) [inline profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php]: Certainly not 'access user profiles'. You seem to have just copy-pasted that from the code a bit higher.

```
case 'user':
  if ($account = user_load($entity_id)) {
    if (user_access('access user profiles')) {
      $render = user_view($account, $view_mode);
       ...
   }
```

File: tests/features/webtools.feature (hunk 10 above)
Comment (cyberwolf) [inline tests/features/webtools.feature]: Ok, fine by me.

File: profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php (hunk 1 above)
Comment (voidtek) [inline profiles/common/modules/custom/nexteuropa_token/src/Entity/ViewModeTokenHandler.php]: I understand now your remark.

**Comment by cyberwolf** [comment]
👍

**Comment by netlooker** [comment]
👍

**Comment by cyberwolf** [comment]
👍

**Comment by netlooker** [comment]
👍

**Pull request merged**